from snco_pipeline.pipeline_utils import (
    annotations_getter, raw_data_getter, results_getter,
    fasta_getter, gtf_getter, barcode_whitelist_getter
)


annotation = annotations_getter(config)
raw_data = raw_data_getter(config)
results = results_getter(config)
get_fasta = fasta_getter(config)
get_gtf = gtf_getter(config)
get_barcode_whitelist = barcode_whitelist_getter(config)
get_conda_env = conda_env_getter(config)

ref_genotypes = set()
genotypes = set()
for ds in config['datasets'].values():
    ref_genotypes.add(ds['reference_genotype'])
    for geno in ds['genotypes'].values():
        genotypes.update(list(geno.values()))


wildcard_constraints:
    dataset_name='|'.join(config['datasets']),
    sample_name="\w+",
    ref='|'.join(ref_genotypes),
    qry='|'.join(genotypes),
    geno='|'.join(genotypes),


tech_types = set([ds['technology'] for ds in config['datasets'].values()])
is_plate_based = tech_types.intersection(['takara_dna', 'plate_wgs'])
is_droplet_based = tech_types.intersection(['10x_rna_v4', '10x_rna_v3', 'bd_rna', '10x_atac'])
if is_plate_based and is_droplet_based:
    raise ValueError('Cannot mix droplet and plate technologies in the same run, please separate to different configs.')


include: "rules/variants.snakefile"
if is_plate_based:
    include: "rules/align_plate.snakefile"
else:
    include: "rules/align_droplet.snakefile"
include: "rules/haplotype.snakefile"


rule all:
    input:
        expand(
            results('haplotypes/{dataset_name}.{file_type}.json'),
            dataset_name=config['datasets'],
            file_type=['markers', 'pred'],
        )